### PR TITLE
Send back invalid transaction data through APIs

### PR DESCRIPTION
**Pull request by delventhalz**
This PR expands the BatchTracker to track invalid Batches, which transaction was rejected, and any data associated with the transaction. This data is then sent back through both the ZMQ and REST APIs with a more robust list of status objects, rather than a simple id to status hashmap.

### CODE REVIEW

File: protos/client.proto [hunk 1]
```diff
@@ -30,16 +30,32 @@ message Leaf {
...
+message BatchStatus {
+    enum Status {
+        COMMITTED = 0;
+        INVALID = 1;
+        PENDING = 2;
+        UNKNOWN = 3;
+    }
+    string batch_id = 1;
+    Status status = 2;
+    string invalid_transaction = 3;
```
Comment (peterschwarz) [inline protos/client.proto]: This possibly should be a list of invalid transactions, but I can't remember if batch processing fails fast.

File: rest_api/sawtooth_rest_api/route_handlers.py [hunk 2]
```diff
@@ -834,6 +834,18 @@ def _set_wait(self, request, validator_query):
                 validator_query.timeout = int(self._timeout * 0.95)
 
     @staticmethod
+    def _format_statuses(statuses):
+        """Reformat converted BatchStatus dicts: drop empty keys, rename 'id'.
```
Comment (peterschwarz) [inline rest_api/sawtooth_rest_api/route_handlers.py]: Perhaps a note that this mutates the status list - returning the value makes it appear like it's creating a new list. Or, just drop the return value, since it looks like it should be properly changed without it.

File: validator/sawtooth_validator/state/batch_tracker.py [hunk 3]
```diff
@@ -93,6 +125,28 @@ def get_statuses(self, batch_ids):
         with self._lock:
             return {b: self.get_status(b) for b in batch_ids}
 
+    def get_invalid_txn_info(self, batch_id):
```
Comment (peterschwarz) [inline validator/sawtooth_validator/state/batch_tracker.py]: Is this used anywhere in this commit?  Or should it be in a different commit?

File: rest_api/sawtooth_rest_api/route_handlers.py [hunk 4]
```diff
@@ -835,16 +835,17 @@ def _set_wait(self, request, validator_query):
 
     @staticmethod
     def _format_statuses(statuses):
-        """Reformat converted BatchStatus dicts: drop empty keys, rename 'id'.
+        """Reformat converted BatchStatus dicts: rename keys and drop empties.
         """
         for status in statuses:
             status['id'] = status.pop('batch_id')
+            status['invalid_message'] = status.pop('message', '')
```
Comment (peterschwarz) [inline rest_api/sawtooth_rest_api/route_handlers.py]: Why rename these?  It makes it so the message is only for invalid statuses, and can't be used for any other status.  No reason that valid transactions could return extended data in the future.

File: validator/sawtooth_validator/state/batch_tracker.py (hunk 3 above)
Comment (delventhalz) [inline validator/sawtooth_validator/state/batch_tracker.py]: Nothing uses any invalid txn functionality in this commit. It just adds the functionality. This method is the only way to access that new functionality. It would only make sense to move it to the next commit (where it is used), if the two commits were combined.

File: rest_api/sawtooth_rest_api/openapi.yaml [hunk 5]
```diff
@@ -125,10 +125,9 @@ paths:
     get:
       summary: Fetches the committed statuses for a set of batches
       description: |
-        Fetches an object with a status for each batch requested. The keys of
-        the data object will be the ids of the batch(es), while the values will
-        be a string status with the values `'COMMITTED'`, `'INVALID'`,
-        `'PENDING'`, or `'UNKNOWN'`.
+        Fetches an array of objects with a status and id for each batch
+        requested. There are for possible statuses with string values
```
Comment (agunde406) [inline rest_api/sawtooth_rest_api/openapi.yaml]: for -> Four

File: protos/client.proto [hunk 6]
```diff
@@ -30,16 +30,32 @@ message Leaf {
...
-//   * COMMITTED - the batch was accepted and has been committed to the chain
-//   * INVALID - the batch failed validation, it should be resubmitted
-//   * PENDING - the batch is still being processed
-//   * UNKNOWN - no status for the batch could be found (possibly invalid)
-enum BatchStatus {
-    COMMITTED = 0;
-    INVALID = 1;
-    PENDING = 2;
-    UNKNOWN = 3;
+// Information about the the status of a batch submitted to the validator.
```
Comment (agunde406) [inline protos/client.proto]: Remove extra "the"

File: protos/client.proto [hunk 7]
```diff
@@ -30,16 +30,32 @@ message Leaf {
...
+//
+// Attributes:
+//     batch_id: The id (header_signature) of the batch
+//     status: The committed status of the batch
+//     invalid_transaction: The header_signature of the txn that was rejected
+//     message: An message with additional information for an invalid batch
```
Comment (agunde406) [inline protos/client.proto]: An -> A
